### DISCUSSION TITLE
export the video with the tracked instances superposed

### DISCUSSION THREAD

**Discussion by ttunja** [talmolab/sleap#937] [Help!] [answered]
I tried to find this functionality through the gui or through the forum. I suspect its burried somewhere in the .slp file which we save after my inference. But my concrete question is:

After performing the inference, how to get the same full video with the colored tracked skeletons superimposed over the animals we wanted to track?

**Comment by roomrys** [comment, marked as answer]
Hi @ttunja,

You can export the video with instances overlaid through the GUI: View > Render Video Clip with Instances...

Let us know if you run into any trouble.

Thanks,
Liezl

**Comment by ttunja** [comment]
Yes I did however every time I try this I get a frozen GUI and the exported video is empty (0 KB). Is there an alternative?

When looking at the console I get 
"
Writing video with 508 frame images...
INFO:sleap.io.visuals:Chunks: 8, chunk size: 64
Exception in thread Thread-3:
Traceback (most recent call last):
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\site-packages\skvideo\io\abstract.py", line 514, in writeFrame
    self._proc.stdin.write(vid.tostring())
BrokenPipeError: [Errno 32] Broken pipe

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\threading.py", line 926, in _bootstrap_inner
    self.run()
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\site-packages\sleap\io\visuals.py", line 131, in writer
    writer_object.add_frame(img, bgr=True)
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\site-packages\sleap\io\videowriter.py", line 108, in add_frame
    self._writer.writeFrame(img)
  File "C:\Users\ttunj\Anaconda3\envs\sleap\lib\site-packages\skvideo\io\abstract.py", line 519, in writeFrame
    raise IOError(msg)
OSError: [Errno 32] Broken pipe

FFMPEG COMMAND:
C:\Users\ttunj\Anaconda3\envs\sleap\Library\bin/ffmpeg.exe -y -r 25 -f rawvideo -pix_fmt rgb24 -s 1272x909 -i - -c:v libx264 -preset superfast -framerate 25 -crf 21 -pix_fmt yuv420p C:\Users\ttunj\OneDrive\Desktop\labels_whole_mice.mp4

FFMPEG STDERR OUTPUT:


"

**Reply by roomrys** [reply]
Hi @ttunja,

Can you try updating to the latest version of SLEAP. There was an issue open about this that should be fixed in the newest release.

Thanks,
Liezl

**Comment by ttunja** [comment]
Hi @roomrys 

thanks this worked to stop the freezing of the sleap GUI, however the file that gets generated afterwards is still empty (0 KB) and cant get played.

**Reply by roomrys** [reply]
@ttunja What type of video are you using (`.mp4`, `.avi`, `.pkg.slp`, directory of images, etc)?

**Comment by ttunja** [comment]
@roomrys input video is .avi, output video (i.e. when I am trying to save it) is obviously  "name".slp.mp4

**Reply by roomrys** [reply]
@ttunja Ok, I quickly tested the video rendering for `avi` on my end and all seemed well - was there any information in your terminal? Is it a multi-video project? Did you select the video to render instances on (by double clicking on the video in the "Video" dialog)? 

If the problem persists, do you mind [opening an issue](https://github.com/talmolab/sleap/issues/new?assignees=&labels=bug&template=bug_report.md)?

**Reply by julia-napoli** [reply]
hi @roomrys, i'm having a similar issue. i am updated to the most updated version of sleap - v1.2.9 at the time of writing - and have successfully tracked an avi video using a model i trained in sleap. now i am trying to render a video with these instances and their skeletons overlaid on it. this is a single-video project and i made sure that i clicked on the single video to render instances on by double clicking it in the dialog. the video saved but is corrupt and is 0 kb. i have attached a screenshot here of the error message in the command line and the error message about the file being corrupt, for reference. thanks for your help in advance!
![Screenshot 2022-12-13 at 2 36 23 PM](https://user-images.githubusercontent.com/73008448/207459743-c34c1a6e-fa11-4c18-897f-799592d6eda5.png)
![Screenshot 2022-12-13 at 2 36 03 PM](https://user-images.githubusercontent.com/73008448/207459739-a83668ef-5839-4ea4-8e05-44889aaf4898.png)

**Comment by roomrys** [comment]
Hi @julia-napoli,

This discussion was continued in the issue #958 where we came to the conclusion that the command we use for rendering videos only works on evenly-sized video. The [fix](https://github.com/talmolab/sleap/issues/958#issuecomment-1249931440) was then to pad the input video with 1 pixel on the odd-sized dimensions using the following command:
```
ffmpeg -i replace_with_input_video_path -c:v libx264 -preset superfast -vf "scale=trunc(iw/2)*2:trunc(ih/2)*2" -crf 23 -pix_fmt yuv420p replace_with_output_video_path
```

Let us know if that helps.

Thanks,
Liezl

**Reply by julia-napoli** [reply]
amazing -- that worked! thank you for your prompt and clear response :)
